### PR TITLE
LRU cache ensureFits() tweaks WRT deadlocking

### PR DESCRIPTION
**Pull request by GiedriusS**
Related to #651. An initial attempt to fix this because I am always nervous that some person could unintentionally make our Thanos Store instance hang even if the index cache is quite big :( 

## Changes

`thanos_store_index_cache_items_overflowed_total` metric added which shows how many times we have tried to add an item to the cache which was simply too big according to the `--index-cache-size` parameter. `ensureFits()` now returns false when we cannot possibly ensure that the passed slice could fit into the cache. The original problem still exists in some form that `setPostings` or `setSeries` could block for a longer time if there are a lot of small items in the cache however at least it shouldn't dead-lock.

## Verification

Wrote unit tests for it and they pass. It is kind of hard to reproduce this unfortunately in a live environment with the changes but I will see what I can do.

### CODE REVIEW

File: pkg/store/cache.go [hunk 1]
```diff
@@ -127,7 +139,11 @@ func (c *indexCache) setPostings(b ulid.ULID, l labels.Label, v []byte) {
 	c.mtx.Lock()
 	defer c.mtx.Unlock()
 
-	c.ensureFits(v)
+	fits := c.ensureFits(v)
+	if !fits {
```
Comment (bwplotka) [inline pkg/store/cache.go]: I think we can inline this.

File: pkg/store/cache.go [hunk 2]
```diff
@@ -115,10 +121,16 @@ func newIndexCache(reg prometheus.Registerer, maxBytes uint64) (*indexCache, err
...
 
-func (c *indexCache) ensureFits(b []byte) {
+// ensureFits tries to make sure that the passed slice will fit into the LRU cache.
+// Returns true if it will fit.
+func (c *indexCache) ensureFits(b []byte) bool {
+	if uint64(len(b)) > c.maxSize {
+		return false
+	}
 	for c.curSize+uint64(len(b)) > c.maxSize {
 		c.lru.RemoveOldest()
```
Comment (bwplotka) [inline pkg/store/cache.go]: I think we could do more aggressive removal if the slice is big enough.

Comment (GiedriusS) [reply pkg/store/cache.go]: Yes but how the implementation would be like? I have two ideas:
* If the slice size exceeds a certain threshold then we could remove 2 or more items at a time
* Get all keys, calculate their sizes, and remove the top X items to make sure there's enough size for the item `b`?

Comment (GiedriusS) [reply pkg/store/cache.go]: Either way, maybe lets leave it for another day? I am mainly interested now in making sure that the loop doesn't go forever

Comment (bwplotka) [reply pkg/store/cache.go]: agree

**Pull request merged**